File: src/main/java/cn/edu/jxnu/base/controller/admin/system/RoleController.java
/* 梦境迷离 (C)2020 */
package cn.edu.jxnu.base.controller.admin.system;

import cn.edu.jxnu.base.controller.BaseController;
import cn.edu.jxnu.base.entity.Role;
import cn.edu.jxnu.base.service.IRoleService;
import cn.edu.jxnu.base.service.IUserService;
import cn.edu.jxnu.base.service.component.MemorandumComponent;
import cn.edu.jxnu.base.service.specification.SimpleSpecificationBuilder;
import cn.edu.jxnu.base.service.specification.SpecificationOperator.Operator;
import cn.edu.jxnu.base.utils.JsonResult;
import javax.servlet.http.HttpServletRequest;
import org.apache.commons.lang3.StringUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.stereotype.Controller;
import org.springframework.ui.ModelMap;
import org.springframework.web.bind.annotation.*;
import reactor.core.publisher.Mono;

/**
 * 系统角色控制类
 *
 * @author 梦境迷离
 * @version V2.0 2020年11月20日
 */
@Controller
@RequestMapping("/admin/role")
public class RoleController extends BaseController {

  @Autowired private IRoleService roleService;

  @Autowired private MemorandumComponent memorandumComponent;

  @Autowired private IUserService userService;

  /**
   * 打开角色管理首页页面
   *
   * @return String
   */
  @RequestMapping(value = {"/", "/index"})
  public String index() {
    return "admin/role/index";
  }

  /**
   * 角色管理分页
   *
   * @param request request
   * @return Mono Page
   */
  @RequestMapping(value = {"/list"})
  @ResponseBody
  public Mono<Page<Role>> list(HttpServletRequest request) {
    SimpleSpecificationBuilder<Role> builder = new SimpleSpecificationBuilder<>();
    String searchText = request.getParameter("searchText");
    if (StringUtils.isNotBlank(searchText)) {
      builder.add("name", Operator.likeAll.name(), searchText);
      builder.addOr("description", Operator.likeAll.name(), searchText);
    }
    return roleService.findAll(builder.generateSpecification(), getPageRequest(request)).log();
  }

  /**
   * 打开添加角色页面
   *
   * @param map map
   * @return String
   */
  @RequestMapping(value = "/add")
  public String add(ModelMap map) {
    Role role = new Role();
    // 传入role实体，提交的时候才有默认值
    map.put("role", role);
    return "admin/role/form";
  }

  /**
   * 打开角色修改页面
   *
   * @param id 角色ID
   * @param map map
   * @return String
   */
  @RequestMapping(value = "/edit/{id}")
  public String edit(@PathVariable Integer id, ModelMap map) {
    Mono<Role> role = roleService.find(id);
    role.log().subscribe(r -> map.put("role", r));
    return "admin/role/form";
  }

  /**
   * 添加或修改角色
   *
   * @param role role
   * @param uCode 用户码
   * @return Mono JsonResult
   */
  @RequestMapping(
      value = {"/edit"},
      method = RequestMethod.POST)
  @ResponseBody
  public Mono<JsonResult> edit(Role role, @RequestParam("uCode") String uCode) {
    try {
      return roleService
          .saveOrUpdate(role)
          .flatMap(
              r ->
                  userService
                      .findByUserCode(uCode)
                      .map(
                          u ->
                              memorandumComponent.saveMemorandum(
                                  uCode,
                                  u.getUserName(),
                                  "修改/新增角色",
                                  role.getRoleKey() + " | " + role.getName())))
          .log()
          .map(r -> JsonResult.success());
    } catch (Exception e) {
      e.printStackTrace();
      return Mono.error(e);
    }
  }

  /**
   * 删除角色
   *
   * @param id 用户ID
   * @param uCode 用户码
   * @return Mono JsonResult
   */
  @RequestMapping(value = "/delete/{id}", method = RequestMethod.POST)
  @ResponseBody
  public Mono<JsonResult> delete(@PathVariable Integer id, @RequestParam("uCode") String uCode) {
    return userService
        .findByUserCode(uCode)
        .flatMap(
            u ->
                roleService
                    .find(id)
                    .flatMap(
                        r -> {
                          memorandumComponent
                              .saveMemorandum(
                                  uCode,
                                  u.getUserName(),
                                  "删除角色",
                                  r.getRoleKey() + " | " + r.getName())
                              .log()
                              .subscribe();
                          return roleService.delete(id);
                        }))
        .log()
        .map(r -> r ? JsonResult.success() : JsonResult.failure("删除失败"))
        .onErrorResume(error -> Mono.just(JsonResult.failure(error.getLocalizedMessage())));
  }

  /**
   * 打开授权页面
   *
   * @param id 角色ID
   * @param map map
   * @return String
   */
  @RequestMapping(value = "/grant/{id}")
  public String grant(@PathVariable Integer id, ModelMap map) {
    Mono<Role> role = roleService.find(id);
    role.log().subscribe(r -> map.put("role", r));
    return "admin/role/grant";
  }

  /**
   * 授权
   *
   * @param id 角色ID
   * @param resourceIds 资源ID
   * @return Mono JsonResult
   */
  @RequestMapping(value = "/grant/{id}", method = RequestMethod.POST)
  @ResponseBody
  public Mono<JsonResult> grant(
      @PathVariable Integer id, @RequestParam(required = false) String[] resourceIds) {
    return roleService
        .grant(id, resourceIds)
        .map(r -> JsonResult.success())
        .log()
        .onErrorResume(error -> Mono.just(JsonResult.failure(error.getLocalizedMessage())));
  }
}

File: src/main/java/cn/edu/jxnu/base/controller/admin/system/RoleController.scala
package cn.edu.jxnu.base.controller.admin.system

import cn.edu.jxnu.base.service.specification.SimpleSpecificationBuilder
import org.springframework.web.bind.annotation.RequestMapping
import cn.edu.jxnu.base.service.IRoleService
import cn.edu.jxnu.base.common.MemorandumUtils
import cn.edu.jxnu.base.service.IUserService
import cn.edu.jxnu.base.controller.BaseController
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.ResponseBody
import org.springframework.data.domain.Page
import cn.edu.jxnu.base.entity.Role
import org.apache.commons.lang3.StringUtils
import cn.edu.jxnu.base.service.specification.SpecificationOperator.Operator
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.ui.ModelMap
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestParam
import cn.edu.jxnu.base.common.JsonResult
import org.springframework.stereotype.Controller

/**
 * 系统角色控制类
 *
 * @author 梦境迷离
 * @time 2018年5月16日
 * @version V2.0
 */
@Controller
@RequestMapping(value = { Array("/admin/role") })
class RoleController @Autowired() (roleService: IRoleService, memorandumUtils: MemorandumUtils, userService: IUserService)
    extends BaseController {

    @RequestMapping(value = { Array("/", "/index") })
    def index() = {
        "admin/role/index"
    }

    @RequestMapping(value = { Array("/list") })
    @ResponseBody
    def list(): Page[Role] = {
        val builder = new SimpleSpecificationBuilder[Role]()
        val searchText = request.getParameter("searchText")
        if (StringUtils.isNotBlank(searchText)) {
            builder.add("name", Operator.likeAll.name(), searchText)
            builder.addOr("description", Operator.likeAll.name(), searchText)
        }
        roleService.findAll(builder.generateSpecification(), getPageRequest())
    }

    @GetMapping(value = Array("/add"))
    def add(map: ModelMap) = {
        val role = new Role()
        map.put("role", role)
        "admin/role/form"
    }

    @GetMapping(value = Array("/edit/{id}"))
    def edit(@PathVariable id: Int, map: ModelMap) = {
        println("role id:" + id)
        val role = roleService.find(id)
        map.put("role", role)
        "admin/role/form"
    }

    @PostMapping(value = { Array("/edit") })
    @ResponseBody
    def edit(role: Role, map: ModelMap, @RequestParam("uCode") uCode: String): JsonResult = {
        try {
            roleService.saveOrUpdate(role)
            memorandumUtils.saveMemorandum(memorandumUtils, uCode, userService.findByUserCode(uCode).getUserName(),
                "修改/新增角色", role.getRoleKey() + " | " + role.getName())
        } catch {
            case e: Exception =>
                return JsonResult.failure(e.getMessage())
        }
        JsonResult.success()
    }

    @PostMapping(value = { Array("/delete/{id}") })
    @ResponseBody
    def delete(@PathVariable id: Int, map: ModelMap, @RequestParam("uCode") uCode: String): JsonResult = {
        try {
            memorandumUtils.saveMemorandum(memorandumUtils, uCode, userService.findByUserCode(uCode).getUserName(),
                "删除角色", roleService.find(id).getRoleKey() + " | " + roleService.find(id).getName())
            roleService.delete(id)
        } catch {
            case e: Exception =>
                return JsonResult.failure(e.getMessage())
        }
        JsonResult.success()
    }

    @GetMapping(value = { Array("/grant/{id}") })
    def grant(@PathVariable id: Int, map: ModelMap) = {
        val role = roleService.find(id)
        map.put("role", role)
        "admin/role/grant"
    }

    @PostMapping(value = { Array("/grant/{id}") })
    @ResponseBody
    def grant(@PathVariable id: Int, @RequestParam(required = false) resourceIds: Array[String],
        map: ModelMap): JsonResult = {
        try {
            roleService.grant(id, resourceIds)
        } catch {
            case e: Exception =>
                e.printStackTrace()
                return JsonResult.failure(e.getMessage())
        }
        JsonResult.success()
    }
}